File: python/handlers/symbol_creator.py
"""
Symbol-creator handlers — create / delete / list / register custom
symbols in .kicad_sym libraries.  Thin wrappers around
`commands.symbol_creator.SymbolCreator`.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any, Dict

from commands.symbol_creator import SymbolCreator

if TYPE_CHECKING:
    from kicad_interface import KiCADInterface

logger = logging.getLogger(__name__)


def handle_create_symbol(iface: "KiCADInterface", params: Dict[str, Any]) -> Dict[str, Any]:
    """Create a new symbol in a .kicad_sym library."""
    logger.info(f"create_symbol: {params.get('name')} in {params.get('libraryPath')}")
    try:
        creator = SymbolCreator()
        return creator.create_symbol(
            library_path=params.get("libraryPath", ""),
            name=params.get("name", ""),
            reference_prefix=params.get("referencePrefix", "U"),
            description=params.get("description", ""),
            keywords=params.get("keywords", ""),
            datasheet=params.get("datasheet", "~"),
            footprint=params.get("footprint", ""),
            in_bom=params.get("inBom", True),
            on_board=params.get("onBoard", True),
            pins=params.get("pins", []),
            rectangles=params.get("rectangles", []),
            polylines=params.get("polylines", []),
            overwrite=params.get("overwrite", False),
        )
    except Exception as e:
        logger.error(f"create_symbol error: {e}")
        return {"success": False, "error": str(e)}


def handle_delete_symbol(iface: "KiCADInterface", params: Dict[str, Any]) -> Dict[str, Any]:
    """Delete a symbol from a .kicad_sym library."""
    logger.info(f"delete_symbol: {params.get('name')} from {params.get('libraryPath')}")
    try:
        creator = SymbolCreator()
        return creator.delete_symbol(
            library_path=params.get("libraryPath", ""),
            name=params.get("name", ""),
        )
    except Exception as e:
        logger.error(f"delete_symbol error: {e}")
        return {"success": False, "error": str(e)}


def handle_list_symbols_in_library(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """List all symbols in a .kicad_sym file."""
    logger.info(f"list_symbols_in_library: {params.get('libraryPath')}")
    try:
        creator = SymbolCreator()
        return creator.list_symbols(
            library_path=params.get("libraryPath", ""),
        )
    except Exception as e:
        logger.error(f"list_symbols_in_library error: {e}")
        return {"success": False, "error": str(e)}


def handle_register_symbol_library(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """Register a .kicad_sym library in KiCAD's sym-lib-table."""
    logger.info(f"register_symbol_library: {params.get('libraryPath')}")
    try:
        creator = SymbolCreator()
        return creator.register_symbol_library(
            library_path=params.get("libraryPath", ""),
            library_name=params.get("libraryName"),
            description=params.get("description", ""),
            scope=params.get("scope", "project"),
            project_path=params.get("projectPath"),
        )
    except Exception as e:
        logger.error(f"register_symbol_library error: {e}")
        return {"success": False, "error": str(e)}

File: src/tools/symbol-creator.ts
/**
 * Symbol creator tools for KiCAD MCP server
 *
 * create_symbol              – add a new symbol to a .kicad_sym library
 * delete_symbol              – remove a symbol from a library
 * list_symbols_in_library    – list all symbols in a .kicad_sym file
 *
 * (Adding a library to the sym-lib-table is handled by the merged
 * register_library tool in library.ts.)
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { CommandFunction, makePassthrough } from "./tool-response.js";

const PinSchema = z.object({
  name: z.string().describe("Pin name, e.g. 'VCC'; '~' for unnamed"),
  number: z.union([z.string(), z.number()]).describe("Pin number, e.g. '1', '2', 'A1'"),
  type: z
    .enum([
      "input",
      "output",
      "bidirectional",
      "tri_state",
      "passive",
      "free",
      "unspecified",
      "power_in",
      "power_out",
      "open_collector",
      "open_emitter",
      "no_connect",
    ])
    .describe("Electrical pin type"),
  at: z
    .object({
      x: z.number().describe("X position in mm"),
      y: z.number().describe("Y position in mm"),
      angle: z
        .number()
        .describe("Direction pin extends FROM body: 0=right, 90=up, 180=left, 270=down"),
    })
    .describe("Pin endpoint position (wire connection point)"),
  length: z.number().optional().describe("Pin length in mm (default 2.54)"),
  shape: z
    .enum([
      "line",
      "inverted",
      "clock",
      "inverted_clock",
      "input_low",
      "clock_low",
      "output_low",
      "falling_edge_clock",
      "non_logic",
    ])
    .optional()
    .describe("Pin graphic shape (default: line)"),
});

const RectSchema = z.object({
  x1: z.number(),
  y1: z.number(),
  x2: z.number(),
  y2: z.number(),
  width: z.number().optional().describe("Stroke width in mm (default 0.254)"),
  fill: z
    .enum(["none", "outline", "background"])
    .optional()
    .describe("Fill type (default: background)"),
});

const PolylineSchema = z.object({
  points: z.array(z.object({ x: z.number(), y: z.number() })).describe("List of XY points in mm"),
  width: z.number().optional().describe("Stroke width in mm (default 0.254)"),
  fill: z.enum(["none", "outline", "background"]).optional(),
});

export function registerSymbolCreatorTools(server: McpServer, callKicadScript: CommandFunction) {
  const passthrough = makePassthrough(callKicadScript);
  // ── create_symbol ────────────────────────────────────────────────────── //
  server.tool(
    "create_symbol",
    "Create a schematic symbol in a .kicad_sym library (file created if missing); run register_library (type=symbol) afterwards so KiCAD finds it. " +
      "Pin positions are wire endpoints. Grid & pin length 2.54 mm; body ±2.54–5.08 mm: " +
      "left pins x=body_left−length angle=0; right x=body_right+length angle=180; " +
      "top y=body_top+length angle=270; bottom y=body_bottom−length angle=90.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file (created if missing)"),
      name: z.string().describe("Symbol name, e.g. 'TMC2209'"),
      referencePrefix: z
        .string()
        .optional()
        .describe("Reference prefix: 'U', 'R', 'J', etc. (default 'U')"),
      description: z.string().optional().describe("Human-readable description"),
      keywords: z.string().optional().describe("Space-separated search keywords"),
      datasheet: z.string().optional().describe("Datasheet URL or '~'"),
      footprint: z
        .string()
        .optional()
        .describe("Default footprint, e.g. 'Package_SO:SOIC-8_3.9x4.9mm_P1.27mm'"),
      inBom: z.boolean().optional().describe("Include in BOM (default true)"),
      onBoard: z.boolean().optional().describe("Include in PCB netlist (default true)"),
      pins: z.array(PinSchema).optional().describe("Pins (empty for graphical-only symbols)"),
      rectangles: z
        .array(RectSchema)
        .optional()
        .describe("Body rectangle(s), typically one for the IC body"),
      polylines: z
        .array(PolylineSchema)
        .optional()
        .describe("Polylines for custom body shapes (op-amp triangles, etc.)"),
      overwrite: z
        .boolean()
        .optional()
        .describe("Replace existing symbol of same name (default false)"),
    },
    passthrough("create_symbol"),
  );

  // ── delete_symbol ────────────────────────────────────────────────────── //
  server.tool(
    "delete_symbol",
    "Remove a symbol from a .kicad_sym library file.",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
      name: z.string().describe("Symbol name to delete"),
    },
    passthrough("delete_symbol"),
  );

  // ── list_symbols_in_library ──────────────────────────────────────────── //
  server.tool(
    "list_symbols_in_library",
    "List symbol names in a .kicad_sym file given its path. Works for unregistered files (e.g. right after create_symbol); for a registered library nickname use list_library_contents (type=symbol).",
    {
      libraryPath: z.string().describe("Path to the .kicad_sym file"),
    },
    passthrough("list_symbols_in_library"),
  );
}
